src/databricks/labs/remorph/reconcile/reconciler/comparator.py

```diff
@@ -0,0 +1,38 @@
+from pyspark.sql import DataFrame
+from pyspark.sql.functions import col, lit, when
+
+from databricks.labs.remorph.reconcile.recon_config import ReconcileOutput
+
+
+class Comparator:
+
+    def compare_data(self, source: DataFrame, target: DataFrame, join_cols: list[str], report: str) -> ReconcileOutput:
+        join = join_cols if report in {"all", "data"} else "hash_value__recon"
+        df = source.alias('s').join(target.alias("t"), join, how='full')
+
+        mismatch = self._get_mismatch_data(df) if report in {"all", "data"} else None
+        missing_in_src = (
+            df.filter(col("s.hash_value__recon").isNull())
+            .select((join_cols if report == "all" else "t.*"))
+            .drop("hash_value__recon")
+        )
+        missing_in_tgt = (
+            df.filter(col("t.hash_value__recon").isNull())
+            .drop("hash_value__recon")
+            .select((join_cols if report == "all" else "s.*"))
+            .drop("hash_value__recon")
+        )
+        return ReconcileOutput(missing_in_src=missing_in_src, missing_in_tgt=missing_in_tgt, mismatch=mismatch)
+
+    @staticmethod
+    def _get_mismatch_data(df: DataFrame) -> DataFrame:
+        return (
+            df.filter((col("s.hash_value__recon").isNotNull()) & (col("t.hash_value__recon").isNotNull()))
+            .withColumn(
+                "hash_match",
+                when(col("s.hash_value__recon") == col("t.hash_value__recon"), lit(True)).otherwise(lit(False)),
```

Comment:
```suggestion
                col("s.hash_value__recon") == col("t.hash_value__recon")
```

Reply:
@vil1 Both Threshold and Data Comparison PR needs to be re-written using sqlglot expressions, ignore till the new changes are in place.

Reply:
Gotcha, although I'm not sure what it has to do with the above suggestion (that involves only the `Dataframe` API, as far as I understand)

Reply:
I meant to save your time, this PR is going to significant changes, much easier for you to review once those changes are in place.

Reply:
Makes sense. So I'll wait until you notify me that this PR is ready for my review. Cheers!